src/content/blog/val-vibes.mdx

```diff
@@ -0,0 +1,166 @@
+
+```
+CREATE TABLE vals_embeddings (id TEXT PRIMARY KEY, embedding VECTOR(1536));
+```
+
+Then we create a [cron val](https://docs.val.town/quickstarts/first-cron/) (or as I like to call them: inter-val…), which we set to run every hour.
+
+First we query all vals, for which <a href="https://www.val.town/u/pomdtr">Achille Lacoin</a> had already made an <a href="https://www.val.town/v/sqlite/db">excellent val</a>:
+
+<figure style={{ maxWidth: "600px" }}>
+  <Image src={QueryVals} alt="" />
+  <figcaption>
+    Querying public vals, by <a href="https://www.val.town/u/pomdtr">Achille Lacoin</a>.
+  </figcaption>
+</figure>
+
+We filter out vals for which we have existing embeddings, and make batches of 100 of the remaining vals.
+
+<figure style={{ maxWidth: "700px" }}>
+  <Image src={Indexing1} alt="" />
+  <figcaption>
+  </figcaption>
+</figure>
+
+Then we fetch embeddings from OpenAI for each batch of 100 in parallel, and save them to Postgres.
+
+<figure style={{ maxWidth: "700px" }}>
+  <Image src={Indexing2} alt="" />
+  <figcaption>
+  </figcaption>
+</figure>
+
+This works! Here are the logs from a recent incremental run:
+
+<figure style={{ maxWidth: "700px" }}>
+  <Image src={Indexing3} alt="" />
+  <figcaption>
+  </figcaption>
+</figure>
+
+Querying is simpler still. We turn the query string into an embedding, and sort by similarity:
+
+<figure style={{ maxWidth: "700px" }}>
+  <Image src={Search1} alt="" />
+  <figcaption>
+  </figcaption>
+</figure>
+
+For a nicer UI, I forked an HTTP val (also by <a href="https://www.val.town/u/pomdtr">Achille Lacoin</a>) and hooked it up to my search function:
+
+<figure style={{ maxWidth: "700px" }}>
+  <Image src={Search2} alt="" />
+  <figcaption>
+  </figcaption>
+</figure>
+
+This works pretty well! Check out all my semantic search prototypes [here](https://www.val.town/v/janpaul123/valtownsemanticsearch).
+
+### Product
+
+We have since integrated this into the search of Val Town properly. The implementation is basically the same: we added a pgvector `embedding` column to our Postgres table that stores vals, and we update the embedding whenever changing the code of a val (or its name or readme).
+
+You can use it by going to the search page and selecting “semantic search”.
+
+<figure style={{ maxWidth: "700px" }}>
+  <Image src={Prod} alt="" />
+  <figcaption>
+  </figcaption>
+</figure>
+
+We're still iterating on this feature. Here are some ideas we have to improve it, but if you have your own, please let us know on [Discord](https://discord.gg/dHv45uN5RY) or [Github](https://github.com/val-town/val-town-product/discussions)!
+
+- Simply concatenate exact and semantic search. Label them differently by highlighting exact matches for exact search, and maybe label semantic results different.
+- For searches that return a lot of exact results, do a bit smarter ranking by using the semantic distance as a ranking feature (in lieu of tf-idf stats with our current search setup).
```

Comment:
```suggestion
- For searches that return a lot of exact results, do a bit smarter ranking by using the semantic distance as a ranking feature, in lieu of tf-idf stats with our current search setup.
```

I don't think this is accurate - our current exact search doesn't do tf-idf.

Reply:
Ah yeah I didn't write this right, I meant in lieu of not having tf-idf stats with our current setup